vochain/scrutinizer/vote.go

```diff
@@ -163,80 +135,125 @@ func (s *Scrutinizer) addLiveVote(envelope *models.Vote) error {
-			if envelope.GetWeight() != nil {
-				weight = new(big.Int).SetBytes(envelope.GetWeight())
-			}
-			if update.Weight == nil {
-				update.Weight = weight
-			} else {
-				update.Weight.Add(update.Weight, weight)
+func (s *Scrutinizer) addProcessToLiveResults(pid []byte) {
+	s.liveResultsProcs.Store(string(pid), true)
+}
+
+func (s *Scrutinizer) delProcessFromLiveResults(pid []byte) {
+	s.liveResultsProcs.Delete(string(pid))
+}
+
+func (s *Scrutinizer) isProcessLiveResults(pid []byte) bool {
+	_, ok := s.liveResultsProcs.Load(string(pid))
+	return ok
+}
+
+// commitVotes adds the votes and weight from results to the local database.
+// It does not overwrite the stored results but update them by adding the new content.
+func (s *Scrutinizer) commitVotes(pid []byte, results *Results) error {
+	s.addVoteLock.Lock()
+	defer s.addVoteLock.Unlock()
+	update := func(record interface{}) error {
+		update, ok := record.(*Results)
+		if !ok {
+			return fmt.Errorf("record isn't the correct type! Wanted Result, got %T", record)
+		}
+		update.Weight.Add(update.Weight, results.Weight)
+
+		if len(results.Votes) > 0 {
+			if len(results.Votes) != len(update.Votes) {
+				return fmt.Errorf("commitVotes: different length for results.Vote and update.Votes")
 			}
-			// Add the vote only if the election is unencrypted
-			if vote != nil {
-				if err := addVote(update.Votes,
-					vote.Votes,
-					weight.Bytes(),
-					p.GetVoteOptions(),
-					p.GetEnvelopeType()); err != nil {
-					return err
+			for i := range update.Votes {
+				if len(update.Votes[i]) != len(results.Votes[i]) {
+					return fmt.Errorf("commitVotes: different number of options for question %d", i)
 				}
-			}
-			// Print some debug info
-			log.Debugf("addVote %v with weight %s on process %x", func() string {
-				if vote == nil {
-					return "encrypted"
+				for j := range update.Votes[i] {
+					update.Votes[i][j].Add(update.Votes[i][j], results.Votes[i][j])
 				}
-				return fmt.Sprintf("%v", vote.Votes)
-			}(), weight.String(), pid)
-			return nil
-		})
-}
+			}
+		}
+		return nil
+	}
 
-func (s *Scrutinizer) computeLiveResults(processID []byte) (*models.ProcessResult, error) {
-	results := &Results{}
-	if err := s.db.FindOne(results, badgerhold.Where("ProcessID").Eq(processID)); err != nil {
-		return nil, err
+	// TODO (PAU):
+	// Since the Commit() callback is executed on a goroutine, if scrutinizer have pending votes to add
+	// and the program is stoped. The missing votes won't be added on next run.
+	// Either a recovery mechanism or a persistent cache should be used in order to avoid this situation.
+
+	var err error
+	// If badger returns "Transaction Conflict" we try again until it works
+	maxTries := 1000
+	for maxTries > 0 {
+		err = s.db.UpdateMatching(&Results{},
+			badgerhold.Where("ProcessID").Eq(pid).And("Final").Eq(false), update)
+		if err == nil {
+			break
+		}
+		if strings.Contains(err.Error(), "Transaction Conflict") {
```

Comment:
use badger.ErrConflict instead: https://pkg.go.dev/github.com/dgraph-io/badger/v2#pkg-variables

If the error is wrapped, then use `errors.Is(err, badger.ErrConflict)`.

Reply:
Mmm, I did not add this error cause I did not want to import a specific KV implementation. So badgerhold can switch from badger v2 to v3 without any change in our side. Do you really think it is worth it? (importing badger v2 just for handling this error).

Reply:
Right, I wasn't thinking about v2 vs v3. If we have good tests which cover the conflict case (e.g. with many reads/writes), then I imagine the tests would catch if they bump to v3 but our error points at v2.

Alternatively, add a short comment to explain the strings.Contains? Most times it's a mistake, so it would be good to clarify why it's wanted here.